examples: Add tcp_sock.py

rdna opened the pull request:
Add example script that goes over all TCP sockets and for every socket
prints:
- TCP state;
- IP:port pair;
- cgroup v2 path.

The script shows basic operations with sockets including getting `struct
sock`, `struct inet_sock`, socket fields and cgroup v2 of socket.

It provides output similar to `ss -nt` or `ss -lnt`:
```
% sudo python36 -m drgn examples/linux/tcp_sock.py | shuf -n 5
TIME_WAIT                                       [::1]:2560                                      [::1]:44041
ESTABLISHED                                     [::1]:54178                                     [::1]:11101  system.slice/async-distillery-server.service
LISTEN                                           [::]:3201                                       [::]:0      system.slice/fbflow.service
ESTABLISHED           [2401:db00:21:718a:face:0:31:0]:2406            [2401:db00:12:a106:face:0:1e:0]:46306  system.slice/configerator_proxy2.service
LISTEN                                        0.0.0.0:11150                                   0.0.0.0:0      system.slice/system-mcrouter.slice/mcrouter@tao.service
```

Since it requires iterating over sockets, necessary `hlist_nulls`
helpers are added as well.

Some functions from `tcp_sock.py` (`sk_fullsock`, `sk_nulls_for_each`,
`inet_sk`) are generally useful and can be moved to a library later.

rdna commented:
[Test failure](https://travis-ci.org/osandov/drgn/jobs/627962419?utm_medium=notification&utm_source=github_status
) seems to be unrelated, `apt-get` failed to install something.
Not sure how to restart it though ..

rdna commented:
In the last pushes: split commonly useful functions to work with nulls lists, network sockets and TCP states into corresponding helper modules.

osandov commented on drgn/helpers/linux/list_nulls.py:
```diff
@@ -0,0 +1,66 @@
+
+def hlist_nulls_for_each_entry(head, type, member):
+    """
+    .. c:function:: hlist_nulls_for_each_entry(type *tpos, struct hlist_nulls_node *pos, struct hlist_nulls_head *head, member)
```
The C signatures are for documenting the helper itself rather than the original C function/macro it came from, so this could be:

```
.. c:function:: hlist_nulls_for_each_entry(type, struct hlist_nulls_head *head, member)
```

On a related note, the names in the C signatures should be consistent with the names in the Python signature for the other helpers.

rdna replied:
Yeah, I mistakenly assumed that it's copy/paste of C signature. Makes sense, I'll cleanup all signatures here and in other places.

osandov commented on drgn/helpers/linux/tcp_states.py:
```diff
@@ -0,0 +1,54 @@
+]
+
+
+class TcpStates(object):
```
Python 3 has the [`enum`](https://docs.python.org/3/library/enum.html) module which we could use here.

osandov commented on examples/linux/tcp_sock.py:
```diff
@@ -0,0 +1,110 @@
+    return '{:>40}:{:<6}'.format(_brackets(ip), port)
+
+
+def _cgroup_path(cgroup):
```
This seems like it could be a useful helper.

rdna replied:
Yep, I have a plan to add a bunch of cgroup related helpers, but wanted to do it in a separate PR when I have cycles + not to make this one too big. LMK if it doesn't work for some reason.

osandov commented on examples/linux/tcp_sock.py:
```diff
@@ -0,0 +1,110 @@
+        # Uncomment to print whole struct:
+        #   print(sk)
+        #   print(inet)
+        #   print(cgroup)
```
The indentation looks too deep here.

rdna replied:
Oops, code move failure, will fix, thanks.

osandov commented on drgn/helpers/linux/tcp_states.py:
```diff
@@ -0,0 +1,54 @@
+    LAST_ACK = 9
+    LISTEN = 10
+    CLOSING = 11
+    NEW_SYN_RECV = 12
```
Are these values guaranteed not to change? If not, then we should grab them from the debug info instead of hard-coding. E.g., something like this:

```python
import enum


def get_tcp_states(prog):
    try:
        return prog.cache['TcpStates']
    except KeyError:
        pass
    # The enum type is anonymous, but we can get it from one of the
    # enumerators.
    tcp_states_type = prog['TCP_ESTABLISHED'].type_.enumerators
    enumerators = [
        (name[4:] if name.startswith('TCP_') else name, value)
        for (name, value) in tcp_states_type
        if name != 'TCP_MAX_STATES'
    ]
    TcpStates = enum.IntEnum('TcpStates', enumerators)
    prog.cache['TcpStates'] = TcpStates
    return TcpStates
```

rdna replied:
Oh, wow.

TIL: that it's possible to get a `enum` from `prog`
TIL: `prog.cache`

These values should never change, since e.g. glibc header `<netinet/tcp.h>` and kernel's `include/uapi/linux/bpf.h` use same enumerations, but new states (even though just one was added since 2005) may appear with delay in new places, e.g. the last enumeration `TCP_NEW_SYN_RECV` was introduced in 2015 but it's not present on my devbox in `<netinet/tcp.h>`. 

The only reason I defined them manually is I didn't know how to get them automatically and getting them automatically is much better IMO since for example it avoids possible problems with "recently" introduced  `TCP_NEW_SYN_RECV`.

I'll switch to your version. Thanks!

rdna replied:
While switching to the automatic enum generation I realized that it would require changing all interfaces that uses TCP states to add additional `prog` argument (e.g. `sk_fullsock`).

That's probably not a big deal, but I'm curious if it's possible to get access to `prog` variable w/o passing it as an argument to helper. The benefit would be have helpers with signatures similar to those in kernel (e.g. same `sk_fullsock`)

osandov replied:
As long as the helper takes a `drgn.Object`, you can grab it from `obj.prog_` (e.g., `sk.prog_` for `sk_fullsock`).

rdna commented:
In the last push:
- consistent order of args in `hlist_nulls_for_each_entry`;
- fixed `c:function` signatures;
- auto generation of TCP states `enum`;
- fixed indentation in the example.

osandov commented:
Thanks, this is great. The Linux kernel helpers are more fine-grained than we need for drgn, so I moved the tcp_states helpers to a generic tcp.py and renamed sock.py to net.py. I also added an `sk_tcpstate` helper and some unit tests. I'll merge this once tests pass.

osandov commented:
Oops, the test failure is because in older kernels, the TCP state definitions don't get compiled in if the kernel isn't compiled with TCP. Easy fix, we shouldn't try to access them until we've created the socket.
